uPortal-webapp/src/main/webapp/WEB-INF/jsp/Invoker/sitemap.jsp

```diff
@@ -0,0 +1,157 @@
+<script src="<rs:resourceURL value="/rs/template/1.0.0/template.js"/>"></script>
+<script
+	src="<rs:resourceURL value="/rs/promise-polyfill/6.0.2/promise.js"/>"></script>
+
+<%-- API URL for fetching layout details to build sitemap --%>
+<c:url value="/api/v4-3/dlm/layout.json" var="layoutApiUrl" />
+
+<%--
+    UI strings for i18n
+--%>
+<spring:message var="i18n_error_loading_sitemap"
+	code="error.loading.sitemap" />
+
+<script language="javascript" type="text/javascript">
+
+fetch('${layoutApiUrl}', {credentials: 'same-origin'})
+    // check for HTTP 2XX Okay response
+    .then(function (response) {
+        if (response.status >= 200 && response.status < 300) {
+            return response;
+        } else {
+            var error = new Error(response.statusText);
+            error.response = response;
+            throw error;
+        }
+    })
+    // extract json from response
+    .then(function (response) {
+        return response.json();
+    })
+    // Generate sitemap
+    .then(function (response) {
+      	sitemapJsonCheck(response, ['layout.navigation.tabs'], "Missing required object path ");
+	
+		// Begin tab row
+       	var tabRowTemplate = document.getElementById('sitemap-tab-row-template');
+		var tabRow = document.importNode(tabRowTemplate.content, true);
+		var tabCounter = 0;
+		_.forEach(response.layout.navigation.tabs, function (tab, tabIndex) {
+			sitemapJsonCheck(tab, ['name','ID','content'], "Missing required object path [layout.navigation.tabs] > ");
+			tabCounter++;
+			// Setup tab link
+			var tabTemplate = document.getElementById('sitemap-tab-template');
+           	var tabHeader = document.importNode(tabTemplate.content, true);
+           	// Add content to tab header template
+           	var tabHeaderLink = tabHeader.querySelector('a');
+           	tabHeaderLink.textContent = tab.name;
+           	tabHeaderLink.href = '/uPortal/f/' + tab.ID + '/normal/render.uP';
+           	var portletList = tabHeader.querySelector('ul');
+		
+			_.forEach(tab.content, function (parentContent, parentContentIndex) {
+				sitemapJsonCheck(parentContent, ['content'], "Missing required object path [layout.navigation.tabs] > content > ");
+
+				_.forEach(parentContent.content, function (portlet, portletIndex) {
+					sitemapJsonCheck(portlet, ['name', 'fname', 'ID'], "Missing required object path [layout.navigation.tabs] > content > content > ");
+
+					// Setup portlet link
+					var portletTemplate = document.getElementById('sitemap-tab-portlet-template');
+           			var portletListItem = document.importNode(portletTemplate.content, true);
+           			// Add content to portlet template
+           			var portletTitle = portletListItem.querySelector('span');
+           			portletTitle.textContent = portlet.name;
+					var portletLink = portletListItem.querySelector('a');
+           			portletLink.href = '/uPortal/f/' + tab.ID + '/p/'+portlet.fname+'.'+portlet.ID+'/max/render.uP';
```

Comment:
I think there's a better way to setup this link.

Reply:
@drewwills are there some JSP/Java helper variables/functions that could help generate this link?

Reply:
I sent something to @cbeach47 in chat.